Allow rdoc-ref to link to non-text files

jeremyevans opened the pull request:
It appears the current behavior of disallowing links to non-text files is deliberate, as it was explicitly added in 3628e196cbbac4767f5b845c5b795c635a671b04.  While the commit message explains the change, it doesn't provide a justification for excluding non-text TopLevels. The issue mentioned in the commit message is also unrelated to the change.

It's just as useful to link to a non-text file as it is to link to a text file, so I think it should be allowed.  I was surprised when it didn't work. I want to use this feature in tilt's documentation.

If there is a reason to disallow it by default, I think this limitation should be documented.  I also I think we should add an option to allow it in that case.

st0012 commented:
I agree with the goal but I think we need a bit more changes to achieve that.

To test the PR, I added `{Link}[rdoc-ref:lib/rdoc/cross_reference.rb]` to `readme.rdoc`. But clicking it doesn't bring me to the right page. The reason seems to be because `cross_reference.rb`'s uri path is `/RDoc/CrossReference.html` but it's not successfully generated in the html link.
And can you add some test cases to `test/rdoc/rdoc_markup_to_html_crossref_test.rb` as well? Thanks

jeremyevans commented:
> I agree with the goal but I think we need a bit more changes to achieve that.
> 
> To test the PR, I added `{Link}[rdoc-ref:lib/rdoc/cross_reference.rb]` to `readme.rdoc`. But clicking it doesn't bring me to the right page. The reason seems to be because `cross_reference.rb`'s uri path is `/RDoc/CrossReference.html` but it's not successfully generated in the html link. 

The link is correct, the issue is that darkfish doesn't generate the page.  Added a commit to fix darkfish, and updated the darkfish tests to test that it is generated. When I was testing earlier versions of this, I was using hanna (as I do for all my projects), which has always generated doc files for non-text files.

> And can you add some test cases to `test/rdoc/rdoc_markup_to_html_crossref_test.rb` as well? Thanks

The first commit already tests resolution of non-text files:

```
  assert_ref @top_level, 'xref_data.rb'
```

Can you let me know which additional tests you would like?

st0012 commented on lib/rdoc/store.rb:
```diff
@@ -723,7 +723,7 @@ def modules_hash
 
   def page(name)
-    @text_files_hash.each_value.find do |file|
+    @files_hash.each_value.find do |file|
```
Can you update the comment too?

st0012 reviewed (commented):
Ok I think I understand the issue here better now:
- Themes can decide whether they want to display non-text files. `hanna` displays them but `darkfish` doesn't
- But RDoc's cross referencing feature doesn't allow referencing to non-text files (e.g. `lib/foo.rb`) regardless if the theme generates them or not

Is my understanding correct?
If it is:
- I think it's better to revert the 2nd commit as it makes darkfish generate a lot of new pages while most of them are empty. Sorry for the back and forth.
- Can you also update `rdoc-ref`'s documentation in `doc/rdoc/markup_reference.rb` to mention how to link to files?

jeremyevans commented:
I removed the darkfish commit, and added commits to updated the comment and show a usage example. Probably best to squash merge these, but I kept them separate to facilitate review.

st0012 reviewed (approved):
Thank you!
